src/Aave.sol

```diff
@@ -0,0 +1,249 @@
+
+  constructor(IAavePool aave, address asset) AaveRouterBase(aave, asset) {}
+
+  fallback() external {
+    uint balance = IERC20(ATOKEN).balanceOf(msg.sender);
+    uint amt = parseAmount(balance, msg.data);
+    IERC20(ATOKEN).safeTransferFrom(msg.sender, address(this), amt);
+
+    // Need to pass MAX_UINT to withdraw full amount after interest accrual.
+    AAVE.withdraw(address(ASSET), amt == balance ? MAX_UINT : amt, msg.sender);
+  }
+}
+
+// =========================
+// ======== Factory ========
+// =========================
+
+contract AaveRouterFactory {
+  IAavePool public immutable AAVE;
+  address public immutable WETH;
+  address public immutable SUPPLY_ETH_ROUTER;
+  address public immutable WITHDRAW_ETH_ROUTER;
+
+  event RoutersDeployed(
+    address supplyRouter, address withdrawRouter, address indexed asset
+  );
+
+  constructor(IAavePool aave, address weth) {
+    AAVE = aave;
+    WETH = weth;
+
+    bytes32 salt = _salt(weth);
+    SUPPLY_ETH_ROUTER = address(new AaveSupplyEth{salt: salt}(aave, weth));
+    WITHDRAW_ETH_ROUTER = address(new AaveWithdrawEth{salt: salt}(aave, weth));
+    emit RoutersDeployed(SUPPLY_ETH_ROUTER, WITHDRAW_ETH_ROUTER, weth);
+  }
+
+  function deploy(address asset) external returns (address, address) {
+    address supplyRouter =
+      address(new AaveSupplyToken{salt: _salt(asset)}(AAVE, asset));
+    address withdrawRouter =
+      address(new AaveWithdrawToken{salt: _salt(asset)}(AAVE, asset));
+
+    emit RoutersDeployed(supplyRouter, withdrawRouter, asset);
+    return (supplyRouter, withdrawRouter);
+  }
+
+  function getRouters(address asset) public view returns (address, address) {
+    if (asset == WETH) {
+      return (SUPPLY_ETH_ROUTER, WITHDRAW_ETH_ROUTER);
+    }
+
+    (address supplyRouter, address withdrawRouter) = computeAddresses(asset);
+    if (supplyRouter.code.length == 0) {
+      return (address(0), address(0));
+    }
+    return (supplyRouter, withdrawRouter);
+  }
+
+  function isDeployed(address asset) external view returns (bool) {
+    (address supplyRouter,) = getRouters(asset);
+    return supplyRouter != address(0);
+  }
+
+  function computeAddresses(address asset)
+    public
+    view
+    returns (address, address)
+  {
+    if (asset == WETH) {
+      return (SUPPLY_ETH_ROUTER, WITHDRAW_ETH_ROUTER);
+    }
+    address supplyRouter = _computeAddress(asset, true);
+    address withdrawRouter = _computeAddress(asset, false);
+    return (supplyRouter, withdrawRouter);
+  }
+
+  function _computeAddress(address asset, bool supply)
+    internal
+    view
+    returns (address)
+  {
+    return Create2.computeCreate2Address(
+      _salt(asset),
+      address(this),
+      supply
+        ? type(AaveSupplyToken).creationCode
+        : type(AaveWithdrawToken).creationCode,
+      abi.encode(AAVE, asset)
+    );
+  }
+
+  function _salt(address asset) internal pure returns (bytes32) {
+    return bytes32(uint(uint160(asset)));
```

Comment:
Should this contain the chain ID?

I was trying to think of how someone could deploy a malicious version of these contracts on other chains, but nothing I could come up with worked. Figured I'd just mention it anyway

Reply:
Agreed, I don't think we need it, these are stateless and permissionless so same address across chains is ok